--- linux-4.12.5-base/drivers/net/wireless/marvell/mwlwifi/mac80211.c	2017-08-10 03:49:38.000000000 +0200
+++ linux/drivers/net/wireless/marvell/mwlwifi/mac80211.c	2017-08-31 21:58:24.395991592 +0200
@@ -286,6 +286,13 @@
 		rc = mwl_fwcmd_set_cdd(hw);
 	}
 
+	if (changed & IEEE80211_CONF_CHANGE_POWER) {
+		rc = mwl_fwcmd_max_tx_power(hw, conf, 0);
+		if (rc)
+			goto out;
+		rc = mwl_fwcmd_tx_power(hw, conf, 0);
+	}
+
 out:
 
 	return rc;
--- linux-4.12.5-base/drivers/net/wireless/marvell/mwlwifi/hif/fwcmd.c	2017-08-10 03:49:38.000000000 +0200
+++ linux/drivers/net/wireless/marvell/mwlwifi/hif/fwcmd.c	2017-08-31 23:26:14.000000000 +0200
@@ -1013,34 +1013,14 @@
 {
 	struct ieee80211_channel *channel = conf->chandef.chan;
 	struct mwl_priv *priv = hw->priv;
-	int reduce_val = 0;
 	u16 band = 0, width = 0, sub_ch = 0;
 	u16 maxtxpow[SYSADPT_TX_POWER_LEVEL_TOTAL];
-	int i, tmp;
+	int i;
 	int rc = 0;
 
 	if (priv->forbidden_setting)
 		return rc;
 
-	switch (fraction) {
-	case 0:
-		reduce_val = 0;    /* Max */
-		break;
-	case 1:
-		reduce_val = 2;    /* 75% -1.25db */
-		break;
-	case 2:
-		reduce_val = 3;    /* 50% -3db */
-		break;
-	case 3:
-		reduce_val = 6;    /* 25% -6db */
-		break;
-	default:
-		/* larger than case 3,  pCmd->MaxPowerLevel is min */
-		reduce_val = 0xff;
-		break;
-	}
-
 	if (channel->band == NL80211_BAND_2GHZ)
 		band = FREQ_BAND_2DOT4GHZ;
 	else if (channel->band == NL80211_BAND_5GHZ)
@@ -1082,16 +1062,11 @@
 		priv->powinited |= MWL_POWER_INIT_1;
 	}
 
-	for (i = 0; i < SYSADPT_TX_POWER_LEVEL_TOTAL; i++) {
-		if (priv->target_powers[i] > priv->max_tx_pow[i])
-			tmp = priv->max_tx_pow[i];
-		else
-			tmp = priv->target_powers[i];
-		maxtxpow[i] = ((tmp - reduce_val) > 0) ? (tmp - reduce_val) : 0;
-	}
+	for (i = 0; i < SYSADPT_TX_POWER_LEVEL_TOTAL; i++)
+		maxtxpow[i] = conf->power_level;
 
 	rc = mwl_fwcmd_set_tx_powers(priv, maxtxpow, HOSTCMD_ACT_GEN_SET,
-				     channel->hw_value, band, width, sub_ch);
+					channel->hw_value, band, width, sub_ch);
 
 	return rc;
 }
@@ -1101,35 +1076,15 @@
 {
 	struct ieee80211_channel *channel = conf->chandef.chan;
 	struct mwl_priv *priv = hw->priv;
-	int reduce_val = 0;
 	u16 band = 0, width = 0, sub_ch = 0;
 	u16 txpow[SYSADPT_TX_POWER_LEVEL_TOTAL];
-	int index, found = 0;
-	int i, tmp;
+	int index;
+	int i;
 	int rc = 0;
 
 	if (priv->forbidden_setting)
 		return rc;
 
-	switch (fraction) {
-	case 0:
-		reduce_val = 0;    /* Max */
-		break;
-	case 1:
-		reduce_val = 2;    /* 75% -1.25db */
-		break;
-	case 2:
-		reduce_val = 3;    /* 50% -3db */
-		break;
-	case 3:
-		reduce_val = 6;    /* 25% -6db */
-		break;
-	default:
-		/* larger than case 3,  pCmd->MaxPowerLevel is min */
-		reduce_val = 0xff;
-		break;
-	}
-
 	if (channel->band == NL80211_BAND_2GHZ)
 		band = FREQ_BAND_2DOT4GHZ;
 	else if (channel->band == NL80211_BAND_5GHZ)
@@ -1173,21 +1128,19 @@
 			priv->cdd = tx_pwr->cdd;
 			priv->txantenna2 = tx_pwr->txantenna2;
 
-			if (tx_pwr->setcap)
+			if (tx_pwr->setcap){
 				priv->powinited = MWL_POWER_INIT_1;
-			else
-				priv->powinited = MWL_POWER_INIT_2;
-
-			for (i = 0; i < SYSADPT_TX_POWER_LEVEL_TOTAL; i++) {
-				if (tx_pwr->setcap)
+				for (i = 0; i < SYSADPT_TX_POWER_LEVEL_TOTAL; i++)
 					priv->max_tx_pow[i] =
 						tx_pwr->tx_power[i];
-				else
+			}
+			else{
+				priv->powinited = MWL_POWER_INIT_2;
+				for (i = 0; i < SYSADPT_TX_POWER_LEVEL_TOTAL; i++)
 					priv->target_powers[i] =
 						tx_pwr->tx_power[i];
 			}
 
-			found = 1;
 			break;
 		}
 	}
@@ -1206,23 +1159,9 @@
 		priv->powinited |= MWL_POWER_INIT_1;
 	}
 
-	for (i = 0; i < SYSADPT_TX_POWER_LEVEL_TOTAL; i++) {
-		if (found) {
-			if ((priv->tx_pwr_tbl[index].setcap) &&
-			    (priv->tx_pwr_tbl[index].tx_power[i] >
-			    priv->max_tx_pow[i]))
-				tmp = priv->max_tx_pow[i];
-			else
-				tmp = priv->tx_pwr_tbl[index].tx_power[i];
-		} else {
-			if (priv->target_powers[i] > priv->max_tx_pow[i])
-				tmp = priv->max_tx_pow[i];
-			else
-				tmp = priv->target_powers[i];
-		}
+	for (i = 0; i < SYSADPT_TX_POWER_LEVEL_TOTAL; i++)
+		txpow[i] = conf->power_level;
 
-		txpow[i] = ((tmp - reduce_val) > 0) ? (tmp - reduce_val) : 0;
-	}
 
 	rc = mwl_fwcmd_set_tx_powers(priv, txpow, HOSTCMD_ACT_GEN_SET_LIST,
 				     channel->hw_value, band, width, sub_ch);
