--- linux/drivers/net/wireless/marvell/Makefile	2016-12-15 17:50:48.000000000 +0100
+++ ./drivers/net/wireless/marvell/Makefile	2016-12-18 19:41:30.020538550 +0100
@@ -4,3 +4,5 @@
 obj-$(CONFIG_MWIFIEX)	+= mwifiex/
 
 obj-$(CONFIG_MWL8K)	+= mwl8k.o
+obj-$(CONFIG_MWLWIFI)   += mwlwifi/
+
--- linux/drivers/net/wireless/marvell/Kconfig	2016-12-15 17:50:48.000000000 +0100
+++ ./drivers/net/wireless/marvell/Kconfig	2016-12-18 19:39:56.284335725 +0100
@@ -14,7 +14,7 @@
 source "drivers/net/wireless/marvell/libertas/Kconfig"
 source "drivers/net/wireless/marvell/libertas_tf/Kconfig"
 source "drivers/net/wireless/marvell/mwifiex/Kconfig"
-
+source "drivers/net/wireless/marvell/mwlwifi/Kconfig"
 config MWL8K
 	tristate "Marvell 88W8xxx PCI/PCIe Wireless support"
 	depends on MAC80211 && PCI
--- linux/firmware/Makefile	2016-01-31 20:29:37.000000000 +0100
+++ ./firmware/Makefile	2016-02-04 19:21:41.239855765 +0100
@@ -39,6 +39,7 @@
 			     bnx2/bnx2-rv2p-09ax-6.0.17.fw \
 			     bnx2/bnx2-mips-06-6.2.1.fw \
 			     bnx2/bnx2-rv2p-06-6.0.15.fw
+fw-shipped-$(CONFIG_MWLWIFI) += mwlwifi/88W8864.bin mwlwifi/88W8897.bin mwlwifi/Marvell_license.txt
 fw-shipped-$(CONFIG_CASSINI) += sun/cassini.bin
 fw-shipped-$(CONFIG_CHELSIO_T3) += cxgb3/t3b_psram-1.1.0.bin \
 				   cxgb3/t3c_psram-1.1.0.bin \
