--- a/drivers/mtd/ofpart.c	2016-09-24 10:10:18.000000000 +0200
+++ a/drivers/mtd/ofpart.c	2016-09-29 22:17:33.000000000 +0200
@@ -25,6 +25,8 @@
 	return of_get_property(pp, "compatible", NULL);
 }
 
+static int mangled_rootblock;
+
 static int parse_ofpart_partitions(struct mtd_info *master,
 				   const struct mtd_partition **pparts,
 				   struct mtd_part_parser_data *data)
@@ -33,6 +35,7 @@
 	struct device_node *mtd_node;
 	struct device_node *ofpart_node;
 	const char *partname;
+	const char *owrtpart = "ubi";
 	struct device_node *pp;
 	int nr_parts, i, ret = 0;
 	bool dedicated = true;
@@ -109,9 +112,15 @@
 		parts[i].offset = of_read_number(reg, a_cells);
 		parts[i].size = of_read_number(reg + a_cells, s_cells);
 
-		partname = of_get_property(pp, "label", &len);
-		if (!partname)
-			partname = of_get_property(pp, "name", &len);
+		if (mangled_rootblock && (i == mangled_rootblock)) {
+			partname = owrtpart;
+		} 
+		else {
+			partname = of_get_property(pp, "label", &len);
+
+			if (!partname)
+				partname = of_get_property(pp, "name", &len);
+		}
 		parts[i].name = partname;
 
 		if (of_get_property(pp, "read-only", &len))
@@ -212,6 +221,18 @@
 	return 0;
 }
 
+static int __init active_root(char *str)
+{
+	get_option(&str, &mangled_rootblock);
+
+	if (!mangled_rootblock)
+		return 1;
+
+	return 1;
+}
+
+__setup("mangled_rootblock=", active_root);
+
 static void __exit ofpart_parser_exit(void)
 {
 	deregister_mtd_parser(&ofpart_parser);
