Index: linux-4.3-rc6/drivers/mtd/nand/pxa3xx_nand.c
===================================================================
--- linux-4.3-rc6.orig/drivers/mtd/nand/pxa3xx_nand.c
+++ linux-4.3-rc6/drivers/mtd/nand/pxa3xx_nand.c
@@ -39,8 +39,8 @@
 
 #include <linux/platform_data/mtd-nand-pxa3xx.h>
 
-#define	CHIP_DELAY_TIMEOUT	msecs_to_jiffies(200)
-#define NAND_STOP_DELAY		msecs_to_jiffies(40)
+#define	CHIP_DELAY_TIMEOUT	msecs_to_jiffies(2000)
+#define NAND_STOP_DELAY		msecs_to_jiffies(400)
 #define PAGE_CHUNK_SIZE		(2048)
 
 /*
